hivemind/moe/client/expert.py

```diff
@@ -89,15 +200,51 @@ def forward(
             for inp, proto in zip(inputs, nested_flatten(info["forward_schema"]))
         ]
 
-        outputs = stub.forward(runtime_pb2.ExpertRequest(uid=ctx.uid, tensors=serialized_tensors))
-
-        deserialized_outputs = [deserialize_torch_tensor(tensor) for tensor in outputs.tensors]
+        size = 0
+        for t in inputs:
+            size += t.element_size() * t.nelement()
+            if size >= DEFAULT_MAX_MSG_SIZE:
+                deserialized_outputs = cls.forward_partial(serialized_tensors, ctx, stub)
+                break
+        else:
+            deserialized_outputs = cls.forward_oneshot(serialized_tensors, ctx, stub)
 
         return tuple(deserialized_outputs)
 
-    @staticmethod
+    @classmethod
+    def forward_partial(cls, serialized_tensors: List[runtime_pb2.Tensor], ctx, stub) -> List[torch.Tensor]:
+        split = [p for t in serialized_tensors for p in split_for_streaming(t, DEFAULT_MAX_MSG_SIZE // 2)]
+
+        outputs = RemoteExpertWorker.run_coroutine(
+            stub.rpc_forward_partial(
+                amap_in_executor(
+                    lambda t: runtime_pb2.ExpertRequest(
+                        uid=ctx.uid,
+                        tensors=[
+                            t,
+                        ],
+                    ),
+                    as_aiter(*split),
+                ),
+            )
+        )
```

Comment:
This looks like black got mad and reformatted the code to be written vertically :) Let's write it in a more readable way: 

```suggestion
        outputs = RemoteExpertWorker.run_coroutine(
            stub.rpc_forward_partial(
                amap_in_executor(
                    lambda t: runtime_pb2.ExpertRequest(uid=ctx.uid, tensors=[t]),
                    as_aiter(*split),
                ),
            )
        )
```

Please also fix the same fragment in `backward_partial()`.

Reply:
To reuse code which decides how to comuncate with server (via streaming or at once) in `RemoteMixtureOfExperts` I made this methods as free functions. Please, review this code once more if it looks OK now or do I have to make some changes